### ISSUE TITLE
[BUG] Some pretty code loads in as blocks

### ISSUE THREAD

**Issue by Vvkmnn** [doomemacs/doomemacs#2083]
**Describe the issue**
Some pretty-code elements won't load in, despite advanced symbols working well. Seems to be a subset of symbols in some languages; haven't experimented if similar types.

![image](https://user-images.githubusercontent.com/1954615/69208437-8fa4f400-0b21-11ea-8010-9d5d453d493c.png)

Current setup, running on Emacs 26.3 via 24efd5b909ccccd5328101577e5afc993866855c:
```elisp
;;; ~/.config/doom/init.el -*- lexical-binding: t; -*-

(doom! :input
       ;;chinese
       ;;japanese

       :completion
       company           ; the ultimate code completion backend
       ;;helm              ; the *other* search engine for love and life
       ;;ido               ; the other *other* search engine...
       ivy               ; a search engine for love and life

       :ui
       ;;deft              ; notational velocity for Emacs
       doom              ; what makes DOOM look the way it does
       doom-dashboard    ; a nifty splash screen for Emacs
       doom-quit         ; DOOM quit-message prompts when you quit Emacs
       fill-column       ; a `fill-column' indicator
       hl-todo           ; highlight TODO/FIXME/NOTE/DEPRECATED/HACK/REVIEW
       hydra
       ;;indent-guides     ; highlighted indent columns
       modeline          ; snazzy, Atom-inspired modeline, plus API
       nav-flash         ; blink the current line after jumping
       ;;neotree           ; a project drawer, like NERDTree for vim
       ophints           ; highlight the region an operation acts on
       (popup            ; tame sudden yet inevitable temporary windows
        +all             ; catch all popups that start with an asterix
        +defaults)       ; default popup rules
       (pretty-code
        +fira)       ; replace bits of code with pretty symbols
       tabs              ; an tab bar for Emacs
       treemacs          ; a project drawer, like neotree but cooler
       unicode           ; extended unicode support for various languages
       vc-gutter         ; vcs diff in the fringe
       vi-tilde-fringe   ; fringe tildes to mark beyond EOB
       window-select     ; visually switch windows
       workspaces        ; tab emulation, persistence & separate workspaces

       :editor
       (evil +everywhere); come to the dark side, we have cookies
       file-templates    ; auto-snippets for empty files
       fold              ; (nigh) universal code folding
       (format +onsave)  ; automated prettiness
       ;;god               ; run Emacs commands without modifier keys
       ;;lispy             ; vim for lisp, for people who don't like vim
       multiple-cursors  ; editing in many places at once
       ;;objed             ; text object editing for the innocent
       parinfer          ; turn lisp into python, sort of
       rotate-text       ; cycle region at point between text candidates
       snippets          ; my elves. They type so I don't have to
       word-wrap         ; soft wrapping with language-aware indent

       :emacs
       dired             ; making dired pretty [functional]
       electric          ; smarter, keyword-based electric-indent
       ibuffer           ; interactive buffer management
       vc                ; version-control and Emacs, sitting in a tree

       :term
       ;;eshell            ; a consistent, cross-platform shell (WIP)
       ;;shell             ; a terminal REPL for Emacs
       ;;term              ; terminals in Emacs
       ;;vterm             ; another terminals in Emacs

       :tools
       ;;ansible
       debugger          ; stepping through code, to help you add bugs
       ;;direnv
       ;;docker
       ;;editorconfig      ; let someone else argue about tabs vs spaces
       ein               ; tame Jupyter notebooks with emacs
       (eval +overlay)     ; run code, run (also, repls)
       flycheck          ; tasing you for every semicolon you forget
       flyspell          ; tasing you for misspelling mispelling
       ;;gist              ; interacting with github gists
       (lookup           ; helps you navigate your code and documentation
        +docsets)        ; ...or in Dash docsets locally
       lsp
       macos             ; MacOS-specific commands
       magit             ; a git porcelain for Emacs
       ;;make              ; run make tasks from Emacs
       ;;pass              ; password manager for nerds
       ;;pdf               ; pdf enhancements
       ;;prodigy           ; FIXME managing external services & code builders
       ;;rgb               ; creating color strings
       ;;terraform         ; infrastructure as code
       ;;tmux              ; an API for interacting with tmux
       ;;upload            ; map local to remote projects via ssh/ftp
       ;;wakatime

       :lang
       ;;agda              ; types of types of types of types...
       ;;assembly          ; assembly for fun or debugging
                                        ; cc
                                        ; +lsp) ; C/C++/Obj-C madness
       ;;clojure           ; java with a lisp
       ;;common-lisp       ; if you've seen one lisp, you've seen them all
       ;;coq               ; proofs-as-programs
       ;;crystal           ; ruby at the speed of c
       ;;csharp            ; unity, .NET, and mono shenanigans
       data              ; config/data formats
                                        ; (elixir
                                        ;   +lsp) ; erlang done right
       ;;elm               ; care for a cup of TEA?
       emacs-lisp        ; drown in parentheses
       ;;erlang            ; an elegant language for a more civilized age
       ;;ess               ; emacs speaks statistics
       ;;faust             ; dsp, but you get to keep your soul
       ;;fsharp           ; ML stands for Microsoft's Language
                                        ; (go ; the hipster dialect
                                        ;   +lsp)
       ;;(haskell +intero) ; a language that's lazier than I am
       ;;hy                ; readability of scheme w/ speed of python
       ;;idris             ;
       ;;(java +meghanada) ; the poster child for carpal tunnel syndrome
       javascript          ; all(hope(abandon(ye(who(enter(here))))))
       ;;julia             ; a better, faster MATLAB
       ;;kotlin            ; a better, slicker Java(Script)
       latex             ; writing papers in Emacs has never been so fun
       ;;lean
       ;;ledger            ; an accounting system in Emacs
       ;;lua               ; one-based indices? one-based indices
       markdown          ; writing docs for people to ignore
       ;;nim               ; python + lisp at the speed of c
       ;;nix               ; I hereby declare "nix geht mehr!"
       ;;ocaml             ; an objective camel
       (org              ; organize your plain life in plain text
        +dragndrop       ; drag & drop files/images into org buffers
        ;;+hugo            ; use Emacs for hugo blogging
        +ipython         ; ipython/jupyter support for babel
        +pandoc          ; export-with-pandoc support
        +pomodoro        ; be fruitful with the tomato technique
        +present)        ; using org-mode for presentations
       ;;perl              ; write code no one else can comprehend
       ;;php               ; perl's insecure younger brother
       ;;plantuml          ; diagrams for confusing people more
       ;;purescript        ; javascript, but functional
       (python             ; beautiful is better than ugly
        +conda             ; snake management
        +lsp)              ; pyls popups
       ;;qt                ; the 'cutest' gui framework ever
       ;;racket            ; a DSL for DSLs
       ;;rest              ; Emacs as a REST client
       ;;rst               ; ReST in peace
       ;; (ruby +lsp) ; 1.step {|i| p "Ruby is #{i.even? ? 'love' : 'life'}"}
       ;;rust              ; Fe2O3.unwrap().unwrap().unwrap().unwrap()
       ;;scala             ; java, but good
       ;;scheme            ; a fully conniving family of lisps
       ;;(sh +lsp) ; she sells {ba,z,fi}sh shells on the C xor
       ;;solidity          ; do you need a blockchain? No.
       ;;swift             ; who asked for emoji variables?
       ;;terra             ; Earth and Moon in alignment for performance.
       (web
        +lsp)               ; the tubes

       :email
       ;;(mu4e +gmail)       ; WIP
       ;;notmuch             ; WIP
       ;;(wanderlust +gmail) ; WIP

       ;; Applications are complex and opinionated modules that transform Emacs
       ;; toward a specific purpose. They may have additional dependencies and
       ;; should be loaded late.
       :app
       ;;calendar
       ;;irc               ; how neckbeards socialize
       ;;(rss +org)        ; emacs as an RSS reader
       ;;twitter           ; twitter client https://twitter.com/vnought
       ;;(write            ; emacs for writers (fiction, notes, papers, etc.)
       ;; +wordnut         ; wordnet (wn) search
       ;; +langtool)       ; a proofreader (grammar/style check) for Emacs

       :config
       ;; For literate config users. This will tangle+compile a config.org
       ;; literate config in your `doom-private-dir' whenever it changes.
       ;;literate

       ;; The default module sets reasonable defaults for Emacs. It also
       ;; provides a Spacemacs-inspired keybinding scheme and a smartparens
       ;; config. Use it as a reference for your own modules.
       (default +bindings +smartparens))
```
```elisp
;;; ~/.config/doom/config.el -*- lexical-binding: t; -*-

;;; ui/pretty-code
;; Fira
(setq   (mac-auto-operator-composition-mode)
        doom-font (font-spec :family "Fira Code" :size 13)
        doom-unicode-font (font-spec :family "Fira Mono" :size 13)
        doom-variable-pitch-font (font-spec :family "Fira Sans" :size 13))
```

Doesn't throw any errors, and in general, Doom seems to be running fine:

<details><pre>
❯❯❯❯ doom doctor                                               ~/.c/doom local 
The doctor will see you now...
> Checking your Emacs version...
> Checking for Emacs config conflicts...
> Checking for private config conflicts...
> Checking for stale elc files...
> Checking Doom Emacs...
  ✓ Initialized Doom Emacs 2.0.9
  ✓ Detected 51 modules
  ✓ Detected 192 packages
  > Checking Doom core for irregularities...
    Found font material-design-icons.ttf
    Found font weathericons.ttf
    Found font octicons.ttf
    Found font fontawesome.ttf
    Found font file-icons.ttf
    Found font all-the-icons.ttf
  > Checking for stale elc files in your DOOMDIR...
  > Checking your enabled modules...

Everything seems fine, happy Emacs'ing!

✓ Finished! (5.8997s)
</pre></details>

It seems to be just an occasional graphical bug; annoying mainly when running into symbols like booleans and strings. 

Tried to solve it via https://github.com/hlissner/doom-emacs/issues/797 and https://github.com/hlissner/doom-emacs/issues/695, but no dice. Double checked I was running the fonts from https://github.com/hlissner/doom-emacs/issues/797#issuecomment-546027790 as well, but I have a screenshot from that experiment later.

Tested with Iosevka through borrowed code from https://github.com/hlissner/doom-emacs/pull/2029 through https://github.com/hlissner/doom-emacs/issues/797#issuecomment-546027790 to be safe (and try new fonts): 

```elisp
;;; init.el -*- lexical-binding: t; -*-

;; Copy this file to ~/.doom.d/init.el or ~/.config/doom/init.el ('doom install'
;; will do this for you). The `doom!' block below controls what modules are
;; enabled and in what order they will be loaded. Remember to run 'doom refresh'
;; after modifying it.
;;
;; More information about these modules (and what flags they support) can be
;; found in modules/README.org.

(doom! :input
       ...
       (pretty-code
        +iosevka)       ; replace bits of code with pretty symbols
       
       (python             ; beautiful is better than ugly
        +conda            ; snake management
        +lsp)                 ; pyls popups
         ...)
```

```elisp
;;; ~/.config/doom/config.el -*- lexical-binding: t; -*-

;;; ui/pretty-code

;;Iosevka
(setq doom-font (font-spec :family "Iosevka" :size 13)
    doom-unicode-font (font-spec :family "Iosevka" :size 13)
    doom-variable-pitch-font (font-spec :family "Iosevka" :size 13))
```

It seems the same boxes in the same places, things like `str` and `bool`, but advanced ligatures continue working perfectly:

![image](https://user-images.githubusercontent.com/1954615/69208304-13121580-0b21-11ea-8b32-b1f7b85d70a9.png)

Weird stuff. Happy to test further. 

Here's what it looks like with Fira Code Symbol for posterity:

```elisp
;;; ~/.config/doom/config.el -*- lexical-binding: t; -*-

;;; ui/pretty-code
;; Fira
(setq   (mac-auto-operator-composition-mode)
        doom-font (font-spec :family "Fira Code" :size 13)
        doom-unicode-font (font-spec :family "Fira Mono" :size 13)
        doom-variable-pitch-font (font-spec :family "Fira Sans" :size 13))
```

![image](https://user-images.githubusercontent.com/1954615/69208890-3fc72c80-0b23-11ea-8e68-a65c6aaab342.png)

Worth noting that some ligatures still work however, which is pretty neat.

<details><pre>
❯❯❯❯ doom info                                                                                                                                                                                            ~/.c/doom local
((emacs
  (version . "26.3")
  (features . "NOTIFY ACL GNUTLS LIBXML2 ZLIB TOOLKIT_SCROLL_BARS THREADS")
  (build . "Nov 19, 2019")
  (buildopts "--enable-locallisppath=/usr/local/share/emacs/site-lisp --infodir=/usr/local/Cellar/emacs-mac/emacs-26.3-z-mac-7.7/share/info/emacs --prefix=/usr/local/Cellar/emacs-mac/emacs-26.3-z-mac-7.7 --with-mac --enable-mac-app=/usr/local/Cellar/emacs-mac/emacs-26.3-z-mac-7.7 --with-gnutls")
  (windowsys . batch)
  (daemonp . server-running))
 (doom
  (version . "2.0.9")
  (build . "HEAD -> develop, origin/develop, origin/HEAD 24efd5b90 2019-11-18 19:10:24 -0500"))
 (system
  (type . darwin)
  (config . "x86_64-apple-darwin19.2.0")
  (shell . "/usr/local/bin/zsh")
  (uname . "Darwin 19.2.0 Darwin Kernel Version 19.2.0: Sat Nov  9 05:51:39 PST 2019; root:xnu-6153.61.1~21/RELEASE_X86_64 x86_64")
  (path "/usr/local/anaconda3/bin" "/usr/local/anaconda3/condabin" "~/.nvm/versions/node/v10.16.0/bin" "~/.google/bin" "~/Documents/dev/go/bin" "/usr/local/Cellar/go/1.13.4/libexec/bin" "/usr/local/opt/llvm/bin" "~/.emacs.d/bin" "/Library/TeX/texbin" "/usr/local/opt/ruby/bin" "/usr/local/anaconda3/bin" "/usr/local/opt/openssl/bin" "/usr/local/sbin" "/usr/local/bin" "/usr/bin" "/bin" "/usr/sbin" "/sbin" "/Applications/VMware Fusion.app/Contents/Public" "/Library/TeX/texbin" "/usr/local/MacGPG2/bin" "/opt/X11/bin" "/Library/Frameworks/Mono.framework/Versions/Current/Commands" "~/Documents/dev/go/bin" "/usr/local/Cellar/go/1.13.4/libexec/bin" "/usr/local/opt/llvm/bin" "~/.emacs.d/bin" "/usr/local/opt/ruby/bin" "/usr/local/anaconda3/bin" "/usr/local/opt/openssl/bin" "/usr/local/sbin" "/Applications/kitty.app/Contents/MacOS" "/usr/local/opt/fzf/bin" "/usr/local/Cellar/emacs-mac/emacs-26.3-z-mac-7.7/libexec/emacs/26.3/x86_64-apple-darwin19.2.0"))
 (config
  (envfile)
  (elc-files . 0)
  (modules :completion company ivy :ui doom doom-dashboard doom-quit fill-column hl-todo hydra modeline nav-flash ophints (popup +all +defaults) (pretty-code +fira) tabs treemacs unicode vc-gutter vi-tilde-fringe window-select workspaces :editor (evil +everywhere) file-templates fold (format +onsave) multiple-cursors parinfer rotate-text snippets word-wrap :emacs dired electric ibuffer vc :tools debugger ein (eval +overlay) flycheck flyspell (lookup +docsets) lsp macos magit :lang data emacs-lisp javascript latex markdown (org +dragndrop +ipython +pandoc +pomodoro +present) (python +conda +lsp) (web +lsp) :config (default +bindings +smartparens))
  (packages "n/a")
  (elpa "n/a")))

</pre></details>

**Comment by Vvkmnn** [comment]
Turns out turning off pretty-code and leaving the font at `Fira Code` with `(mac-auto-operator-composition-mode)` seems to keep some ligatures while dodging the blocks. Temporary workaround:


![image](https://user-images.githubusercontent.com/1954615/69214035-d56ab800-0b33-11ea-8fc7-d0573afcef18.png)

**Comment by hlissner** [comment]
Are you still experiencing this issue?

**Comment by pgronkievitz** [comment]
> Are you still experiencing this issue?

Yeah, at least for me there are _some_ broken characters. It appears that default `True` symbol is broken for me with default configuration. I haven't found anything else yet

**Comment by xf-zhao** [comment]
The exact same thing happens to me, too.

**Comment by arnfaldur** [comment]
I have had this issue for a long time.
Using JetBrains Mono it applies to the true, bool, and list ligatures and switching to Liberation Mono for Powerline makes true, false, str, bool, and list render as blocks.
Interestingly, I can copy the incorrectly rendering text from emacs into a terminal and they render properly there. The terminal uses Liberation Mono for Powerline.

**Comment by tshearman** [comment]
+1

**Comment by pradyungn** [comment]
Still having this issue. Anyone found a workaround yet?

**Comment by rabdulatipoff** [comment]
Same issue as the previous users, yet with a twist: Iosevka seems to handle every character that I usually end up having well, while other fonts (Fira, Hasklig, patched Fantasque Sans Mono NF) have this subset of substitution characters displaying as blocks with pretty-mode enabled. Tried setting my own set of characters for the ligatures module and it works at times, I didn't yet figure out why sometimes my config  is ignored.

**Comment by tko** [comment]
FWIW I think this may have been due to choice of fonts all along. Even after the PR I was seeing plain blocks rather than fancy F and T for false and true. After going through lots of fonts I figured using
```(setq doom-symbol-font (font-spec :family "STIXGeneral"))```
makes it work for me (macos)

Found the font using browser devtools on the false/true symbols on [ligatures/README](https://github.com/doomemacs/doomemacs/blob/master/modules/ui/ligatures/README.org) -- the README was rendering fine in browser, not so much in emacs, before the fonts change.
